main/lsp/wrapper.h

```diff
@@ -3,32 +3,45 @@
 
 #include "spdlog/spdlog.h"
 // has to come before the next spdlog include. This comment stops formatter from reordering them
-#include "lsp.h"
 #include "main/lsp/LSPMessage.h"
+#include "main/lsp/lsp.h"
 #include "spdlog/sinks/stdout_color_sinks.h"
 #include <string_view>
 namespace sorbet::realmain::lsp {
 
-class LSPOutputToVector;
+class LSPOutput;
+class LSPProgrammaticInput;
 class LSPConfiguration;
+
 class LSPWrapper final {
-    static const std::string EMPTY_STRING;
+    /** The LSP 'server', which runs in the same thread as LSPWrapper (unless multithreading is enabled) */
+    const std::shared_ptr<LSPLoop> lspLoop;
+    const std::shared_ptr<LSPConfiguration> config_;
+    const std::shared_ptr<LSPOutput> output;
 
-    /** The LSP 'server', which runs in the same thread as LSPWrapper. */
-    std::unique_ptr<LSPLoop> lspLoop;
+    /** If multithreading is enabled, contains the thread running LSPLoop. */
+    std::unique_ptr<Joinable> lspThread;
+    /** If multithreading mode is enabled, contains the input object used to feed messages to lspThread. */
+    const std::shared_ptr<LSPProgrammaticInput> input;
 
     /**
-     * Sorbet assumes we 'own' this object; keep it alive to avoid memory errors.
+     * Sorbet assumes we 'own' the following three objects; keep them alive to avoid memory errors.
      */
-    std::unique_ptr<WorkerPool> workers;
-    std::shared_ptr<spd::sinks::ansicolor_stderr_sink_mt> stderrColorSink;
-    std::shared_ptr<spd::logger> typeErrorsConsole;
-    std::shared_ptr<LSPConfiguration> config_;
-    std::shared_ptr<LSPOutputToVector> output;
-
-    /** Contains shared constructor logic. */
-    void instantiate(std::unique_ptr<core::GlobalState> gs, const std::shared_ptr<spdlog::logger> &logger,
-                     bool disableFastPath);
+    const std::unique_ptr<WorkerPool> workers;
+    const std::shared_ptr<spd::sinks::ansicolor_stderr_sink_mt> stderrColorSink;
+    const std::shared_ptr<spd::logger> typeErrorsConsole;
+
+    static std::unique_ptr<LSPWrapper>
+    createInternal(std::unique_ptr<core::GlobalState> gs, std::shared_ptr<options::Options> options,
+                   const std::shared_ptr<spdlog::logger> &logger, bool disableFastPath,
+                   std::optional<std::function<void(std::unique_ptr<LSPMessage>)>> &&processResponse,
+                   std::shared_ptr<spd::sinks::ansicolor_stderr_sink_mt> stderrColorSink = nullptr,
+                   std::shared_ptr<spd::logger> typeErrorsConsole = nullptr);
+
+    static std::unique_ptr<LSPWrapper>
+    createInternal(std::string_view rootPath, std::shared_ptr<options::Options> options,
+                   std::optional<std::function<void(std::unique_ptr<LSPMessage>)>> &&processResponse,
+                   int numWorkerThreads, bool disableFastPath);
```

Comment:
It's not a big deal in any way, but is there a reason these have to have the same name? Given their arguments, they don't seem like they're meaningfully overloaded versions of each other; it might be nice to be able to clearly disambiguate them by name instead of just by argument structure.

Reply:
No longer a problem in the recent changes.